kitsune/users/views.py

```diff
@@ -290,17 +290,19 @@ def edit_contribution_area(request):
 @require_http_methods(["GET", "POST"])
 def edit_watch_list(request):
     """Edit watch list"""
-    watches = Watch.objects.filter(user=request.user).order_by("content_type", "id")
-
-    watch_list = []
-    for item in watches:
-        if item.content_object is not None:
-            if item.content_type.name == "question":
-                # Only list questions that are not archived
-                if not item.content_object.is_archived:
-                    watch_list.append(item)
-            else:
-                watch_list.append(item)
+    watches = (
+        Watch.objects.filter(user=request.user)
+        .select_related("content_type")
+        .prefetch_related("content_object")
+        .order_by("content_type", "id")
+    )
```

Comment:
Probably is more efficient to exclude from fetching items that you don't care about.

```python
watches = (
    Watch.objects.filter(
        user=request.user
    )
    .exclude(content_object__isnull=True) 
    .exclude(Q(content_type__model="question") & Q(content_object__is_archived=True))
    .select_related("content_type")
    .prefetch_related("content_object")
    .order_by("content_type", "id")
)

watch_list = paginate(request, watches)
```

Reply:
the reason for the Q object is to ensure that it will filter out `is_archived` attributes only for questions. Otherwise it will potentially raise AttributeError for content_types that don't have that attribute defined in their model

Reply:
That's a good catch. Thanks @akatsoulas

Reply:
The only requested optimization I could keep was adding the is_null exclusion and removing the later check. I also noticed a potential issue with the POST content making multiple saves vs. a single bulk update, and altered that.